pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,463 @@
+	}
+
+	if _, _, err := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(progressing),
+		v1helpers.UpdateConditionFn(available),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *OLMOperatorRemovalController) markFinished(message string) error {
+	progressing := operatorapi.OperatorCondition{
+		Type:    c.Name() + operatorapi.OperatorStatusTypeProgressing,
+		Reason:  "Finished",
+		Status:  operatorapi.ConditionFalse,
+		Message: message,
+	}
+	available := operatorapi.OperatorCondition{
+		Type:    c.Name() + operatorapi.OperatorStatusTypeAvailable,
+		Reason:  "Finished",
+		Status:  operatorapi.ConditionTrue,
+		Message: message,
+	}
+
+	// Clear the old namespace annotation - the driver has been fully removed.
+	if err := c.saveOperatorNamespace(""); err != nil {
+		return err
+	}
+	_, _, err := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(progressing),
+		v1helpers.UpdateConditionFn(available),
+	)
+	return err
+}
+
+func (c *OLMOperatorRemovalController) ensureOperatorDeploymentRemoved(namespace, name string) (bool, error) {
+	// Do not actively remove the Deployment here, that's OLM's job.
+	_, err := c.deploymentLister.Deployments(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	klog.V(4).Infof("Deleted Deployment %s/%s", namespace, name)
+	// Don't report the Deployment is removed, wait until IsNotFound error above
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) ensureDriverDeploymentRemoved(ctx context.Context, namespace, name string) (bool, error) {
+	_, err := c.deploymentLister.Deployments(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}
+	err = c.kubeClient.AppsV1().Deployments(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		return false, err
+	}
+	klog.V(4).Infof("Deleted driver Deployment %s/%s", namespace, name)
+	// Don't report the Deployment is removed, wait until IsNotFound error above
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) ensureDriverDaemonSetRemoved(ctx context.Context, namespace, name string) (bool, error) {
+	_, err := c.daemonSetLister.DaemonSets(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}
+	err = c.kubeClient.AppsV1().DaemonSets(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		return false, err
+	}
+	klog.V(4).Infof("Deleted driver DaemonSet %s/%s", namespace, name)
+	// Don't report the DaemonSet is removed, wait until IsNotFound error above
+	return false, nil
+}
+
+func (c *OLMOperatorRemovalController) ensureCRRemoved(ctx context.Context, res schema.GroupVersionResource) (bool, error) {
+	cr, err := c.dynamicClient.Resource(res).Get(ctx, oldCRName, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
```

Comment:
`apierrors` and `errors`: you might be importing the same package twice

Reply:
good catch! using `apierrors`.